### PR TITLE
Issue43 instructor

### PR DESCRIPTION
**Pull request by goldlilys**
When I run gulp test:unit:tdd , I'm getting one error:

```
Chrome 40.0.2214 (Mac OS X 10.10.1) LOG: 'test Backbone'
.......
Chrome 40.0.2214 (Mac OS X 10.10.1) Instructor model  when updating the model  does not save when instructor is empty  FAILED
	Expected spy Invalid to have been called with [ { cid: 'c25', attributes: { id: 1 }, _changing: false, _previousAttributes: {  }, changed: {  }, id: 1, _pending: false, _events: { change: [ { callback: Function, context: undefined, ctx: <circular reference: Object> } ], invalid: [ { callback: spy on Invalid, context: undefined, ctx: <circular reference: Object> } ] }, validationError: 'firstName cannot be empty' }, 'instructor cannot be empty', { validate: true, validationError: 'instructor cannot be empty' } ] but actual calls were [ { cid: 'c25', attributes: { id: 1 }, _changing: false, _previousAttributes: {  }, changed: {  }, id: 1, _pending: false, _events: { change: [ { callback: Function, context
![screenshot 2015-01-27 12 37 05](https://cloud.githubusercontent.com/assets/702616/5926639/6f50b154-a621-11e4-938c-9b24d0d953ad.png)
: undefined, ctx: <circular reference: Object> } ], invalid: [ { callback: spy on Invalid, context: undefined, ctx: <circular reference: Object> } ] }, validationError: 'firstName cannot be empty' }, 'firstName cannot be empty', { validate: true, validationError: 'firstName cannot be empty' } ].
	    at Object.<anonymous> (/var/folders/g7/f6s075793fg7b9spwxnrqp_w0000gn/T/b3df38910feefcb1b73756431b10360219e783ad.browserify:246:24 <- client/spa/js/instructor/spec/instructor.model.spec.js:36:0)
................
Chrome 40.0.2214 (Mac OS X 10.10.1): Executed 24 of 29 (1 FAILED) (skipped 3) (0.09 secs / 0.039 secs)
```

It's coming from instructor.model.spec.js, but have no idea why it's going there when I'm not updating the model.

Another question is in the instructor.view.spec.js, how do I create new instances of Model, and not run into this other error:

```
[12:31:37] Starting 'test:api'...
FF.....

Failures:
1) Frisby Test: Create an instructor with post
	[ POST http://localhost:3000/api/instructors/ ]
1.1) Expected { firstName: 'Tom', lastName: 'Shell', skills: 'Scripting', id: 3 } to contain json { firstName: 'Jeff', lastName: 'Thomas', skills: 'C++' }, false.

2) Frisby Test: Create an instructor with post
	[ POST http://localhost:3000/api/instructors/ ]
2.1) Expected { firstName: 'Emily', lastName: 'Row', skills: 'Diving', id: 4 } to contain json { firstName: 'Jeff', lastName: 'Thomas', skills: 'C++' }, false.

7 specs, 2 failures
```

it's a frisby problem, but the Expected part is only comparing the values with the first instance of the model, and not itself.

Other commented out parts is trying to create the edit button... still working on that.

### CODE REVIEW

**Comment by jadedsurfer** [comment]
You have multiple validations coded for your model. The first validation checked is for firstName not instructor. When the test fails it tells you what the spy actually got for arguments ("but actual calls were"). The primary difference between the expectation you have in the test and the actual arguments is the text of the error. "instructor cannot be empty"

**Comment by goldlilys** [comment]
After much looking at the code closely, I changed the validations on the instructor.model.js to be instructor instead of checking the properties and everything passes. Yay. But now a new question, where do we put validations for the properties like the firstName, and lastName? Does that go inside the 
```
xit('initializes with custom logic', function(){
    });
```

**Comment by treyhunner** [comment]
Is this ready for code review?

**Comment by goldlilys** [comment]
Yep

**Comment by jeffreytu** [comment]
When I view one id, let's say /1, click the edit button and save, then go to a different id /2, the edit and delete button stop working. I think you may need to put a view.remove in the controller.

**Comment by goldlilys** [comment]
@jeffreytu I'm not sure what you mean by view.remove in the controller. Is that similar to model.destroy in the view ? Not sure where you would place that in the controller?

**Comment by jeffreytu** [comment]
This is good to merge now. All tests pass. :+1:

**Comment by Rauld1** [comment]
I reviewed and it looks good to me too.  Passing tests as well. :+1:

File: client/spa/js/instructor/instructor.controller.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+var Backbone = require('../vendor/index').Backbone;
+var $ = require('../vendor/index').$;
+var Model = require('./instructor.model');
+var View = require('./instructor.view');
+
+module.exports = Backbone.Controller.extend({
+  routes: {
+    'instructor/:id': 'showInstructor'
```
Comment (tbashor) [inline client/spa/js/instructor/instructor.controller.js]: The convention for the route should be instructors/:id. Use the plural to same consistent with the REST api.

File: client/spa/js/instructor/spec/instructor.model.spec.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+describe('Instructor model ', function(){
+  var model;
+
+  describe('when creating a new model ', function(){
+    beforeEach(function(){
+      model = new Model();
+    });
+
+    xit('initializes with custom logic', function(){
+    });
```
Comment (tbashor) [inline client/spa/js/instructor/spec/instructor.model.spec.js]: Looks like in the model's initialize you are listening to a 'change' event. You may want to put that in the test or eliminate that listener.

File: client/spa/js/instructor/spec/instructor.model.spec.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  });
+
+  describe('when updating the model for instructor ', function(){
+    var errorSpy;
+
+    beforeEach(function(){
```
Comment (tbashor) [inline client/spa/js/instructor/spec/instructor.model.spec.js]: Try this:

beforeEach(function(){
      errorSpy = jasmine.createSpy('Invalid');
      model = new Model({
        id: 1,
        firstName: 'foo',
        lastName: 'foo',
        skills: 'foo'
      });
      model.on('invalid', errorSpy);
    });

    it('does not save when firstName is empty ', function(){
      model.set('firstName', null);
      model.save();
      expect(errorSpy).toHaveBeenCalled();
      expect(errorSpy.calls.mostRecent().args[0]).toBe(model);
      expect(errorSpy.calls.mostRecent().args[1][0]).toEqual(
        'firstName cannot be empty');
    });

    it('does not save when lastName is empty ', function(){
      model.set('lastName', null);
      model.save();
      expect(errorSpy).toHaveBeenCalled();
      expect(errorSpy.calls.mostRecent().args[0]).toBe(model);
      expect(errorSpy.calls.mostRecent().args[1][0]).toEqual(
        'lastName cannot be empty');
    });

    it('does not save when skills is empty ', function(){
      model.set('skills', null);
      model.save();
      expect(errorSpy).toHaveBeenCalled();
      expect(errorSpy.calls.mostRecent().args[0]).toBe(model);
      expect(errorSpy.calls.mostRecent().args[1][0]).toEqual(
        'skills cannot be empty');
    });

File: client/spa/js/instructor/instructor.model.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    lastName: '',
+    skills: ''
+  },
+  urlRoot: '/api/instructors',
+  initialize: function(){
+    this.on('change', function(){
+      this.trigger('foo', 'bar');
+    });
+  },
+  validate: function(attrs){
```
Comment (tbashor) [inline client/spa/js/instructor/instructor.model.js]: Try this:

var errors = [];
    if (!attrs.firstName){
      errors.push('firstName cannot be empty');
    }
    if (!attrs.lastName){
      errors.push('lastName cannot be empty');
    }
    if (!attrs.skills){
      errors.push('skills cannot be empty');
    }
    return errors;

File: client/spa/js/instructor/spec/instructor.model.spec.js (hunk 3 above)
Comment (tbashor) [inline client/spa/js/instructor/spec/instructor.model.spec.js]: This is even better:

  describe('when updating the model for instructor ', function(){


    beforeEach(function(){
    
      model = new Model({
        id: 1,
        firstName: 'foo',
        lastName: 'foo',
        skills: 'foo'
      });
  
    });

    it('does not save when firstName is empty ', function(){
      model.set('firstName', null);
      model.save();
      expect(model.validationError).toEqual(['firstName cannot be empty']);
    });

**Comment by jeffreytu** [comment]
When I view an instructor id, I don't see the edit button anymore. Must be a conflict with our class names when my branch was merged with master. I have the same class name of edit, and you have `<button class="edit">Edit</button>` in your template. We can either prefix our class names like i-edit for yours and lr-edit for mine. Or you can use my template if you like to have the same class names.

File: client/spa/js/instructor/instructor.html [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+}
+.success {
+  color: green;
+}
+</style>
+<section id="instructor">
+  <div class="main container">
+    <h1><%- firstName %> <%- lastName %></h1>
+    <p><%- skills %></p>
+    <button class="edit">Edit</button>
```
Comment (jeffreytu) [inline client/spa/js/instructor/instructor.html]: When I view an instructor id, I don't see the edit button anymore. Must be a conflict with our class names when my branch was merged with master. I have the same class name of edit, and you have <button class="edit">Edit</button> in your template. We can either prefix our class names like i-edit for yours and lr-edit for mine. Or you can use my template if you like to have the same class names.

Comment (goldlilys) [inline client/spa/js/instructor/instructor.html]: Ok, I'll add a different class for my templates so it doesn't overwrite yours and lessen the # of lines.

**Comment by goldlilys** [comment]
@jeffreytu  Alright, try it again and see if everything is passing and all the buttons are displaying correctly for you.

**Comment by tbashor** [comment]
Looks like enough people have looked this over and tests are passing so I'll merge this.

**Pull request merged**